Order by tuple

devcrafter opened the pull request:
### Changelog category (leave one):
- Performance Improvement

Fixes #38772

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
A less efficient execution plan can be generated for query with ORDER BY (a, b) than for ORDER BY a, b

devcrafter committed:
Fix: ORDER BY with braces inefficient execution

devcrafter committed:
Fix: clear children in order by element during rewriting

- remove unnecessary code for this PR

devcrafter committed:
Fix: rewrite only outer braces

devcrafter commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into order_by_with_braces

yakov-olkhovskiy reviewed (approved):
LGTM
consider maybe unnecessary changes

yakov-olkhovskiy commented:
BTW this optimization makes WITH FILL works for such case 👍

devcrafter commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -886,25 +886,27 @@ static FillColumnDescription getWithFillDescription(const ASTOrderByElement & or
 }
 
-SortDescription InterpreterSelectQuery::getSortDescription(const ASTSelectQuery & query, ContextPtr context_)
+SortDescription InterpreterSelectQuery::getSortDescription(const ASTSelectQuery & query, const ContextPtr & context_)
```
Copying `ContextPtr` leads to counter increment, which is not necessary. There is a follow-up regarding this in #38902. I just touched the code in first commits of the fix

devcrafter committed:
Tests

+ check that EXPLAIN SYNTAX return the same result for ordinary ORDER BY and ORDER BY tuple
+ performance

devcrafter commented:
@Mergifyio update

devcrafter committed:
Merge branch 'master' into order_by_with_braces

devcrafter committed:
Fix: remove heeavy performance tests, introduced within this PR

devcrafter committed:
Merge remote-tracking branch 'origin/master' into order_by_with_braces

devcrafter committed:
Simplify test

- complexity comes from dreaming about EXPLAIN queries as an ordinary
  ones returning scalar result. So they can be used in SQL tests
  instead of shell ones

novikd commented on tests/queries/0_stateless/02353_order_by_tuple.sh:
```diff
@@ -0,0 +1,21 @@
+# check that both queries have the same AST after rewrite, EXPLAIN SYNTAX returns it in form of query
+##################
+QUERY_ORDER_BY="SELECT number AS a, number % 2 AS b FROM numbers(10) ORDER BY a DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1, b DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1"
+QUERY_ORDER_BY_TUPLE="SELECT number AS a, number % 2 AS b FROM numbers(10) ORDER BY (a, b) DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1"
```
@devcrafter I'm not sure if we can do such an optimization. Is the query without optimization and with this `WITH FILL` valid?
